feat: enhance linting configuration and improve Makefile validation

jaxom-widby-nttd opened the pull request:
Improved the linting process by adding comprehensive configuration options for golangci-lint, including timeout settings. Updated Makefiles to enhance validation and help targets, ensuring better handling of arguments. Introduced a pre-commit configuration for consistency and updated the .gitignore file to include common temporary files. Additionally, updated dependencies in Makefiles and ensured the NOTICE file reflects the current year.

jaxom-widby-nttd committed:
feat: add comprehensive .gitignore and pre-commit configuration; update Makefiles for improved validation and help targets; update Makefiles for better handling of order of arguments (ala golangci-lint https://github.com/launchbynttdata/launch-workflows/issues/50 ); updated handling of environment variables; dependency checks in Makefile(s); updated versions in files that are being brought forward during execution stages (linkfiles/*); pre-commit added for consistency; NOTICE updated to current year

jaxom-widby-nttd committed:
Update to handling for timeout of golangci-lint as variables may not be exposed in a way that the environment variables can be well overridden currently. This logic intends to determine which is the highest value in minutes or hours and provide this as an override if required, Environment Variable, Default Value and .golangci.yaml are evaluated for this timeout variable.

Updated documentation

jaxom-widby-nttd committed:
feat: Updated base pre-commit-config to match closely with the deployed (in the linkfiles directory) version (without the terraform or go checks since neither are present. Added .secrets-baseline into root and linkefiles directories for inclusion in updates (future) for https://github.com/launchbynttdata/launch-common-automation-framework/pull/37 (and as it is/was required in the .pre-commit-config.yaml being deployed as a part of this repo for terraform modules.

chris11-taylor-nttd commented on .pre-commit-config.yaml:
```diff
@@ -0,0 +1,60 @@
+  #       entry: golangci-lint run --fix
+  #       types: [go]
+  #       language: golang
+  #       pass_filenames: false
```
```suggestion
```

Let's just remove the commented blocks if we're not using them.

jaxom-widby-nttd commented on linkfiles/.secrets.baseline:
This is required for proposed changes in https://github.com/launchbynttdata/launch-common-automation-framework/pull/37

jaxom-widby-nttd replied:
After offline discussions I have opted to remove this as the baseline should be created by the repo maintainer during the build of the repo

jaxom-widby-nttd committed:
Remove commented out sections

Signed-off-by: jaxom-widby-nttd <jaxom.widby@nttdata.com>

jaxom-widby-nttd commented on linkfiles/.tflint.hcl:
```diff
@@ -1,13 +1,4 @@
-
-  varfile = []
-  variables = []
+plugin "terraform" {
+  enabled = true
+  preset  = "recommended"
```
Mirroring https://github.com/launchbynttdata/tf-aws-module-template/pull/1

ben-vaughan-nttd commented on Makefile:
```diff
@@ -10,81 +10,139 @@
 REPO_MANIFEST ?= manifests/terraform_modules/seed/manifest.xml
 
-# Settings to pull in Nexient version of (google) repo utility that supports environment substitution:
+# Settings to pull in standard version of (google) repo utility that supports environment substitution
```
This is pulling in a Launch-modified version of the repo tool... not the "standard" version of the tool.

jaxom-widby-nttd committed:
Delete linkfiles/.secrets.baseline

Removing as the repo maintainer/author should be running "detect-secrets scan > .secrets.baseline" during the process rather than working off of this as a template

Signed-off-by: jaxom-widby-nttd <jaxom.widby@nttdata.com>

jaxom-widby-nttd commented on tasks/golang/Makefile:
```diff
@@ -1,45 +1,202 @@
+	else \
+		echo "$(GO_LINT_TIMEOUT)"; \
+	fi \
+)
 endef
 
-# Check for Go files. If they exist, run tests. Either runs only readonly tests(default) or tests except readonly ones
-define go_test
-	$(FIND) $(1)/ -name '*.go' |$(GREP) $(2) $(GO_TEST_READONLY_DIRECTORY) | $(GREP) -q '\.go' || exit 0; $(GO) test -v -count=1 -timeout=$(GO_TEST_TIMEOUT) $$($(GO) list ./$(1)/...|$(GREP) $(2)  $(GO_TEST_READONLY_DIRECTORY)) ;
+# Runs golangci-lint on a directory if Go files exist
+# Uses only configurable variables - no hard-coded values
+# Smart timeout resolution:
+#   - If config file exists with timeout >= GO_LINT_TIMEOUT: use config file (no --timeout flag)
+#   - If config file exists but GO_LINT_TIMEOUT > config timeout: override with --timeout flag
+#   - If no config file or no timeout in config: use --timeout flag with GO_LINT_TIMEOUT
+# Arguments:
+#   $(1) - Directory to lint (e.g., "tests")
+# https://golangci-lint.run/docs/configuration/ notes that configuration file variables are overridden by CLI flags
+define go_lint
+	@if $(FIND) $(1)/ -name '*.go' 2>/dev/null | $(GREP) -q '\.go'; then \
+		echo "Running golangci-lint on ./$(1)/..."; \
+		OVERRIDE_TIMEOUT="$(call should_override_timeout,$(GO_LINT_CONFIG))"; \
+		if [ -n "$$OVERRIDE_TIMEOUT" ]; then \
+			echo "Overriding config timeout with: $$OVERRIDE_TIMEOUT"; \
+		elif [ -f "$(GO_LINT_CONFIG)" ]; then \
+			CONFIG_TIMEOUT=$$(grep -E '^[[:space:]]*timeout:' "$(GO_LINT_CONFIG)" | sed 's/.*timeout:[[:space:]]*//' | tr -d ' "'\''\`' 2>/dev/null); \
+			if [ -n "$$CONFIG_TIMEOUT" ]; then \
+				echo "Using config file timeout: $$CONFIG_TIMEOUT"; \
+			fi; \
+		fi; \
+		$(GOLANGCI_LINT) run \
+			$$(test -n "$$OVERRIDE_TIMEOUT" && echo "--timeout $$OVERRIDE_TIMEOUT") \
+			$$(test -f "$(GO_LINT_CONFIG)" && echo "-c $(GO_LINT_CONFIG)") \
+			$(GO_LINT_VERBOSE) \
+			$(GO_LINT_EXTRA_FLAGS) \
+			./$(1)/...; \
+	else \
+		echo "No Go files found in $(1)/, skipping lint"; \
+	fi
+endef
```
Validated the logic of this through the following iterations:

1. Pre setup:
    a. Environment variable not set
    b. .golangci.yaml timeout = 15m
    c. Makefile timeout = 15m
2. Run "make go/lint", result:
  - PASS - 15m
3. Adjust .golangci.yaml timeout to 16m and run test:
  - PASS - 16m
4. Adjust Makefile default to 17m and run test:
  - PASS - 17m
5. Set environment variable to 18m and run test:
  - PASS - 18m
6. Set Enviroment variable to 2h and run test:
  - PASS - 2h
7. Set environment variable to blank:
  - PASS - 17m (fell back to default in Makefile)
8. Set .golangci.yaml to 4h:
  - PASS - 4h
9. Rename .golangci.yaml to .temp:
  - PASS - 17m (default in Makefile)
10. Comment out Line setting environment variable with .golangci.yaml in the wrong name:
  - PASS - Uses default value for command at 5 min IIRC, Doesnt report the value in terminal output but does execute still
11. Set all back and still runs as:
  - PASS - 15 min
